// @hubspot/local-dev-lib/constants/config.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HUBSPOT_ACCOUNT_TYPE_STRINGS = exports.HUBSPOT_ACCOUNT_TYPES = exports.MIN_HTTP_TIMEOUT = exports.DEFAULT_PORTAL = exports.DEFAULT_ACCOUNT = exports.ALLOW_USAGE_TRACKING = exports.HTTP_USE_LOCALHOST = exports.ENV = exports.HTTP_TIMEOUT = exports.DEFAULT_CMS_PUBLISH_MODE = exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND = exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID = exports.DEFAULT_ACCOUNT_OVERRIDE_FILE_NAME = exports.GLOBAL_CONFIG_PATH = exports.HUBSPOT_CONFIGURATION_FILE = exports.HUBSPOT_CONFIGURATION_FOLDER = exports.ARCHIVED_HUBSPOT_CONFIG_YAML_FILE_NAME = exports.DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME = void 0;
const lang_1 = require("../utils/lang");
const path_1 = __importDefault(require("path"));
const os_1 = __importDefault(require("os"));
exports.DEFAULT_HUBSPOT_CONFIG_YAML_FILE_NAME = 'hubspot.config.yml';
exports.ARCHIVED_HUBSPOT_CONFIG_YAML_FILE_NAME = 'archived.hubspot.config.yml';
exports.HUBSPOT_CONFIGURATION_FOLDER = '.hscli';
exports.HUBSPOT_CONFIGURATION_FILE = 'config.yml';
exports.GLOBAL_CONFIG_PATH = path_1.default.join(os_1.default.homedir(), exports.HUBSPOT_CONFIGURATION_FOLDER, exports.HUBSPOT_CONFIGURATION_FILE);
exports.DEFAULT_ACCOUNT_OVERRIDE_FILE_NAME = '.hsaccount';
exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID = 'DEFAULT_ACCOUNT_OVERRIDE_ERROR_INVALID_ID';
exports.DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND = 'DEFAULT_ACCOUNT_OVERRIDE_ERROR_ACCOUNT_NOT_FOUND';
exports.DEFAULT_CMS_PUBLISH_MODE = 'defaultCmsPublishMode';
exports.HTTP_TIMEOUT = 'httpTimeout';
exports.ENV = 'env';
exports.HTTP_USE_LOCALHOST = 'httpUseLocalhost';
exports.ALLOW_USAGE_TRACKING = 'allowUsageTracking';
exports.DEFAULT_ACCOUNT = 'defaultAccount';
exports.DEFAULT_PORTAL = 'defaultPortal';
exports.MIN_HTTP_TIMEOUT = 3000;
exports.HUBSPOT_ACCOUNT_TYPES = {
    DEVELOPMENT_SANDBOX: 'DEVELOPMENT_SANDBOX',
    DEVELOPER_TEST: 'DEVELOPER_TEST',
    APP_DEVELOPER: 'APP_DEVELOPER',
    STANDARD_SANDBOX: 'STANDARD_SANDBOX',
    STANDARD: 'STANDARD',
};
exports.HUBSPOT_ACCOUNT_TYPE_STRINGS = {
    DEVELOPMENT_SANDBOX: (0, lang_1.i18n)('lib.accountTypes.developmentSandbox'),
    STANDARD_SANDBOX: (0, lang_1.i18n)('lib.accountTypes.standardSandbox'),
    DEVELOPER_TEST: (0, lang_1.i18n)('lib.accountTypes.developerTest'),
    APP_DEVELOPER: (0, lang_1.i18n)('lib.accountTypes.appDeveloper'),
    STANDARD: (0, lang_1.i18n)('lib.accountTypes.standard'),
};
