// @hubspot/cms-components/dist/chunk-4KWQ7KCR.js
// src/config/buildStaticEnvConfig.ts
function buildStaticEnvConfig(portalID, hublet, hsEnv, hsDeployed, hsJSRGates, hsRenderSource) {
  const getPortalID = () => portalID;
  const getHubID = getPortalID;
  const getHublet = () => hublet;
  const getInternalHSEnv = () => hsEnv;
  const getIsDeployed = () => hsDeployed;
  const getInternalRenderSource = () => hsRenderSource;
  const getGates = () => hsJSRGates;
  const toJSON = () => ({
    portalID,
    hublet,
    hsEnv,
    hsDeployed,
    hsJSRGates
  });
  const equals = (envConfig) => {
    return portalID === envConfig.getPortalID() && hsEnv === envConfig.getInternalHSEnv() && hublet === envConfig.getHublet() && hsDeployed === envConfig.getIsDeployed();
  };
  return {
    getPortalID,
    getHubID,
    getHublet,
    getInternalHSEnv,
    getIsDeployed,
    equals,
    getGates,
    getInternalRenderSource,
    toJSON
  };
}

// src/config/singleton.ts
var hsConfigMapKey = typeof Symbol !== "undefined" ? Symbol.for("hs-config-map-singleton") : "HS_CONFIG_MAP_SINGLETON";
var hsConfigKey = typeof Symbol !== "undefined" ? Symbol.for("hs-config-singleton") : "HS_CONFIG_SINGLETON";
function ensureConfigMap() {
  if (!window[hsConfigMapKey]) {
    window[hsConfigMapKey] = /* @__PURE__ */ new Map();
  }
  return window[hsConfigMapKey];
}
function initConfigSingleton(portalID, hublet, hsEnv, hsDeployed, hsJSRGates) {
  if (import.meta.env.SSR) {
    throw new Error(
      "Config singletons only work in the browser, not server-side"
    );
  }
  const envConfig = buildStaticEnvConfig(
    portalID,
    hublet,
    hsEnv,
    hsDeployed,
    hsJSRGates
  );
  const configMap = ensureConfigMap();
  const existingPortalConfig = configMap.get(portalID);
  if (existingPortalConfig) {
    if (envConfig.equals(existingPortalConfig)) {
      return;
    }
    console.warn(
      `Updating environment config for portal ${portalID}. Old config: ${JSON.stringify(
        existingPortalConfig.toJSON()
      )}, new config: ${JSON.stringify(envConfig.toJSON())}`
    );
  }
  configMap.set(portalID, envConfig);
  if (!window[hsConfigKey] || configMap.size === 1) {
    window[hsConfigKey] = envConfig;
  }
}
function internalGetConfigSingletonForTest(portalID) {
  if (portalID) {
    const configMap = window[hsConfigMapKey];
    return configMap == null ? void 0 : configMap.get(portalID);
  }
  return window[hsConfigKey];
}
function internalUpdateConfigSingletonForTest(environmentConfig, portalID) {
  if (portalID) {
    const configMap = ensureConfigMap();
    configMap.set(portalID, environmentConfig);
  }
  window[hsConfigKey] = environmentConfig;
}
function initConfigSingletonFromJSON({
  portalID,
  hublet,
  hsEnv,
  hsDeployed,
  hsJSRGates
}) {
  return initConfigSingleton(portalID, hublet, hsEnv, hsDeployed, hsJSRGates);
}
function assertConfigSingletonDefined(portalID) {
  if (portalID) {
    const configMap = window[hsConfigMapKey];
    if (!configMap || !configMap.has(portalID)) {
      throw new Error(`No environment config singleton has been setup for portal ${portalID}`);
    }
  } else if (!window[hsConfigKey]) {
    throw new Error("No environment config singleton has been setup");
  }
}
function getConfigSingleton(portalID) {
  if (import.meta.env.SSR) {
    throw new Error(
      "Config singletons only work in the browser, not server-side"
    );
  }
  assertConfigSingletonDefined(portalID);
  if (portalID) {
    const configMap = window[hsConfigMapKey];
    return configMap.get(portalID);
  }
  return window[hsConfigKey];
}

export {
  buildStaticEnvConfig,
  initConfigSingleton,
  internalGetConfigSingletonForTest,
  internalUpdateConfigSingletonForTest,
  initConfigSingletonFromJSON,
  getConfigSingleton
};
